### PR TITLE
Put metadata and schemas in dict representations

### LINKED ISSUE #686: Metadata schemas and top-level metadata not included in `asdict` and `fromdict`
I'm pretty sure they should be as other wise they will silently disappear in some workflows.

### PR DESCRIPTION
**Pull request by benjeffery**
Fixes #686 
This is backward compatible with existing dict representations as schemas (and top-level metadata and schema) are not required on `fromdict` and if empty are not included in `asdict`.
Note that this also adds a `metadata_schema` argument to `set_columns`.

### CODE REVIEW

**Comment by benjeffery** [comment]
I'm missing some coverage, don't merge for now.

**Comment by benjeffery** [comment]
@jeromekelleher This is now fully-backward compatible with older `asdict` and `fromdict`. Should be good to go.

File: python/tests/test_dict_encoding.py [hunk 1]
```diff
@@ -404,3 +487,27 @@ def test_provenances(self):
             "provenances",
             ["record", "record_offset", "timestamp", "timestamp_offset"],
         )
+
+    def test_top_level_metadata(self):
+        tables = get_example_tables()
+        d = tables.asdict()
+        d["metadata"] = None
```
Comment (jeromekelleher) [inline python/tests/test_dict_encoding.py:494]: We also want to test what happens when we ``del d["metadata"]``, right?

File: python/tests/test_dict_encoding.py [hunk 2]
```diff
@@ -404,3 +487,27 @@ def test_provenances(self):
...
+        self.assertNotIn("metadata", out)
+        tables = tskit.TableCollection.fromdict(out)
+        # We only removed the metadata, not the schema. So empty bytefield
+        # still gets interpreted
+        self.assertEqual(tables.metadata, {"top-level": []})
+
+    def test_top_level_metadata_schema(self):
+        tables = get_example_tables()
+        d = tables.asdict()
+        d["metadata_schema"] = None
```
Comment (jeromekelleher) [inline python/tests/test_dict_encoding.py:507]: Likewise here

File: python/tests/test_tables.py [hunk 3]
```diff
@@ -1913,6 +1947,13 @@ def test_from_dict(self):
         t2 = tskit.TableCollection.fromdict(d)
         self.assertEquals(t1, t2)
 
+    def test_roundtrip_dict(self):
+        ts = msprime.simulate(10, mutation_rate=1, random_seed=1)
+        t1 = ts.tables
+        self.add_metadata(t1)
+        t2 = tskit.TableCollection.fromdict(t1.asdict())
+        self.assertEqual(t1, t2)
```
Comment (jeromekelleher) [inline python/tests/test_tables.py:1955]: Can we add another test here to make sure that we also round-trip without adding the metadata?

File: python/tskit/tables.py [hunk 4]
```diff
@@ -1635,7 +1669,9 @@ def append_columns(self, metadata=None, metadata_offset=None):
         :type metadata_offset: numpy.ndarray, dtype=np.uint32.
         """
         self.ll_table.append_columns(
-            dict(metadata=metadata, metadata_offset=metadata_offset)
+            dict(
```
Comment (jeromekelleher) [inline python/tskit/tables.py:1672]: we should be able to just leave ``metadata_schema`` out of this, right?

File: python/tskit/tables.py [hunk 5]
```diff
@@ -1995,6 +2035,19 @@ def __setstate__(self, state):
     @classmethod
     def fromdict(self, tables_dict):
         tables = TableCollection(tables_dict["sequence_length"])
+        # These try-except can be removed once msprime etc are returning dicts with
```
Comment (jeromekelleher) [inline python/tskit/tables.py:2038]: I think we'll always have metadata and metadata_schema optional in the dict encoding, so this check will be permanent.

File: python/_tskitmodule.c [hunk 6]
```diff
@@ -1977,6 +2178,18 @@ write_table_arrays(tsk_table_collection_t *tables, PyObject *dict)
...
         }
+        if (table_descs[j].metadata_schema_length > 0) {
+            array = make_Py_Unicode_FromStringAndLength(table_descs[j].metadata_schema,
+                table_descs[j].metadata_schema_length);
+            if (array == NULL) {
+                goto out;
+            }
+            if (PyDict_SetItemString(table_dict, "metadata_schema", array) != 0) {
+                goto out;
+            }
```
[the remark this answers is not in the document]
Comment (benjeffery) [reply python/_tskitmodule.c:2189]: There was a leak, this was the cause. Fixed in https://github.com/tskit-dev/tskit/pull/687/commits/bb530cad2fc4b7269249304fee13102dd4fe3287

Not sure how I did that one - I remember removing it for some reason. Brain has been absent last couple of days!

File: python/_tskitmodule.c [hunk 7]
```diff
@@ -2013,6 +2226,30 @@ dump_tables_dict(tsk_table_collection_t *tables)
     Py_DECREF(val);
     val = NULL;
 
+    if (tables->metadata_schema_length > 0) {
+        val = make_Py_Unicode_FromStringAndLength(tables->metadata_schema, tables->metadata_schema_length);
```
[the remark this answers is not in the document]
Comment (benjeffery) [reply python/_tskitmodule.c]: https://github.com/tskit-dev/tskit/pull/687/commits/00d82b25da4276bd7dae024219384270f6cc840d

File: python/tests/test_dict_encoding.py [hunk 8]
```diff
@@ -114,6 +145,32 @@ class TestRoundTrip(unittest.TestCase):
     """
 
     def verify(self, tables):
+        tables.metadata_schema = tskit.MetadataSchema(
```
[the remark this answers is not in the document]
Comment (benjeffery) [reply python/tests/test_dict_encoding.py]: https://github.com/tskit-dev/tskit/pull/687/commits/a16864974a018ecb6125ab0d05328567d90ae2fd

File: python/tests/test_dict_encoding.py [hunk 9]
```diff
@@ -173,7 +230,7 @@ def test_missing_sequence_length(self):
     def test_missing_tables(self):
         tables = get_example_tables()
         d = tables.asdict()
-        table_names = set(d.keys()) - {"sequence_length"}
+        table_names = d.keys()
```
[the remark this answers is not in the document]
Comment (benjeffery) [reply python/tests/test_dict_encoding.py]: https://github.com/tskit-dev/tskit/pull/687/commits/fe03c04f546227e6936004b92307ce091a6317f3

File: python/tests/test_dict_encoding.py [hunk 10]
```diff
@@ -304,22 +361,26 @@ def verify_optional_column(self, tables, table_len, table_name, col_name):
             )
         )
 
-    def verify_offset_pair(self, tables, table_len, table_name, col_name):
+    def verify_offset_pair(
```
[the remark this answers is not in the document]
Comment (benjeffery) [reply python/tests/test_dict_encoding.py:382]: Population table didn't call this method - and so it's columns where not checked. For it to be able  to call it I had to skip the tests that fail as it is the only table where metadata is required.

File: python/tests/test_dict_encoding.py (hunk 1 above)
Comment (benjeffery) [reply python/tests/test_dict_encoding.py:513]: https://github.com/tskit-dev/tskit/pull/687/commits/e59142657d1e9a27d15f39e46ed18d52bef98d9d

File: python/tests/test_dict_encoding.py (hunk 2 above)
Comment (benjeffery) [reply python/tests/test_dict_encoding.py:533]: https://github.com/tskit-dev/tskit/pull/687/commits/e59142657d1e9a27d15f39e46ed18d52bef98d9d

File: python/tests/test_tables.py (hunk 3 above)
Comment (benjeffery) [reply python/tests/test_tables.py:1958]: https://github.com/tskit-dev/tskit/pull/687/commits/57edc3d76982290667c5039ff42cf4f88743e7db

File: python/tskit/tables.py (hunk 4 above)
Comment (benjeffery) [reply python/tskit/tables.py:1672]: We can't as this is calling the low-level method.

File: python/tskit/tables.py (hunk 5 above)
Comment (benjeffery) [reply python/tskit/tables.py]: https://github.com/tskit-dev/tskit/pull/687/commits/482e1f2b1046a521549e9ea3a9adc1d599fe9aa1

**Comment by benjeffery** [comment]
@jeromekelleher Thanks, think I have addressed them all. Sorry there were quite a few!

File: python/tests/test_dict_encoding.py (hunk 1 above)
Comment (jeromekelleher) [reply python/tests/test_dict_encoding.py:513]: I don't think this is what we want. The dict encoding is the de-facto interchange format, and so we must be robust to reading in data from older versions of msprime and tsinfer (etc). Erroring out if there's an optional key missing is something we just can't do. The semantics are pretty straightforward (and already handled in the maps-to-None case), so I think it's an easy fix. All we need to do is update ``get_table_dict_value`` to return None when the key is missing and required is false.

I just tried it out, and it seems to work fine. There's a little bit of breakage, but should only be a few minutes to clear up.

This will also simplify some of the functions I pointed out below, where we're adding in keys mapping to None when the parameters are missing.

File: python/tests/test_lowlevel.py [hunk 11]
```diff
@@ -1939,6 +1939,7 @@ def f(mutations):
                     ancestral_state_offset=ancestral_state_offset,
                     metadata=None,
                     metadata_offset=None,
+                    metadata_schema=None,
```
Comment (jeromekelleher) [inline python/tests/test_lowlevel.py:1942]: Shouldn't need this

File: python/tskit/tables.py (hunk 4 above)
Comment (jeromekelleher) [reply python/tskit/tables.py:1672]: Should be OK now, right?

File: python/tests/test_dict_encoding.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (benjeffery) [reply python/tests/test_dict_encoding.py:382]: The boolean parameter only controls the test for the optionality of the columns. The rest of the function still runs, testing that leaving out one of the pair raises an error and that invalid offsets raises an error. Previously to this PR the population table was missing these tests.

[the remark this answers is not in the document]
Comment (jeromekelleher) [reply python/tests/test_dict_encoding.py:382]: Ahhh, sorry, my bad. Hard to parse diffs out of context sometimes.

File: python/tests/test_dict_encoding.py (hunk 1 above)
Comment (benjeffery) [reply python/tests/test_dict_encoding.py:541]: I've has a go in https://github.com/tskit-dev/tskit/pull/687/commits/682b97da754cb07b00969d8d75c6b06d0314976f Missing used to be a `ValueError` and `None` a `TypeError` I've had to harmonise this to `TypeError` we hit other exceptions that have `TypeError` such as those for a missing pair in a ragged set when letting `None` fall through.

File: python/tests/test_lowlevel.py (hunk 11 above)
Comment (benjeffery) [reply python/tests/test_lowlevel.py]: Yep gone in https://github.com/tskit-dev/tskit/pull/687/commits/682b97da754cb07b00969d8d75c6b06d0314976f

File: python/tskit/tables.py (hunk 4 above)
Comment (benjeffery) [reply python/tskit/tables.py:1667]: Yep gone in https://github.com/tskit-dev/tskit/pull/687/commits/682b97da754cb07b00969d8d75c6b06d0314976f

**Comment by benjeffery** [comment]
I've added a version in https://github.com/tskit-dev/tskit/pull/687/commits/37f2bd669e19ad901992d4c852c59fd82098b5d6. I've used 1.1 as 1.0 is the verison that had no version. In your comment you mention setting a default to 1.0 - I think this would go in code that is reading back the version, which we're not doing yet as we're not using it.

**Comment by benjeffery** [comment]
I've added a hardcoded msprime example - but I can't see where in tsinfer it is making dicts. A search for "asdict" fails in the tsinfer source.

File: python/tests/test_dict_encoding.py [hunk 12]
```diff
@@ -404,3 +557,51 @@ def test_provenances(self):
...
+        tables = tskit.TableCollection.fromdict(out)
+        self.assertEqual(str(tables.metadata_schema), "")
+        # Missing is tested in TestMissingData above
+
+
+class TestExamples(unittest.TestCase):
+    def test_pickled_examples(self):
+        seen_current_msprime = False
+        test_dir = pathlib.Path(__file__).parent / "data/dict-encodings"
+        for filename in os.listdir(test_dir):
```
Comment (jeromekelleher) [inline python/tests/test_dict_encoding.py:594]: Might as well use pathlib:
```
for filename in test_dir.glob("*.pkl"):
      # etc
```

File: python/tests/test_dict_encoding.py [hunk 13]
```diff
@@ -404,3 +557,51 @@ def test_provenances(self):
...
+                "msprime" in filename
+                and filename.split("-")[1] == f"{msprime.__version__}.pkl"
+            ):
+                seen_current_msprime = True
+            with open(test_dir / filename, "rb") as f:
+                d = pickle.load(f)
+                lwt = c_module.LightweightTableCollection()
+                lwt.fromdict(d)
+                tskit.TableCollection.fromdict(d)
+        self.assertTrue(seen_current_msprime)
```
Comment (jeromekelleher) [inline python/tests/test_dict_encoding.py:607]: This is going to break for some developers who install development versions of msprime. I don't think it gains us enough to make the hassle worth it, so I'd say just read in what's there and maybe assert that at least one file has been read (disturbingly easy for tests like this to fall through without doing anything).

Comment (jeromekelleher) [reply python/tests/test_dict_encoding.py:607]: Also, we'd have to update this for every release of msprime - hardly seems worth it, since the dict encoding is going to change extremely rarely.

File: python/tests/test_dict_encoding.py (hunk 12 above)
Comment (benjeffery) [reply python/tests/test_dict_encoding.py]: Fixed in https://github.com/tskit-dev/tskit/pull/687/commits/a8d83651c0842dacfbb4605ca067c82efbec1831

File: python/tests/test_dict_encoding.py (hunk 13 above)
Comment (benjeffery) [reply python/tests/test_dict_encoding.py]: Yeah, this was wrong! Fixed in https://github.com/tskit-dev/tskit/pull/687/commits/a8d83651c0842dacfbb4605ca067c82efbec1831

**Comment by benjeffery** [comment]
@jeromekelleher Ok, let me know if good for a squash.

**Comment by jeromekelleher** [comment]
I'll give one more look over once #687 is merged @benjeffery, but I'd be surprised if anything else shows up at this point.

**Comment by benjeffery** [comment]
#687 is this PR?

**Comment by jeromekelleher** [comment]
> #687 is this PR?

:facepalm:
